ACPs/77-reinventing-subnets/README.md

```diff
@@ -46,77 +46,154 @@ For standalone networks like the Avalanche Primary Network, this is done by conn
-    // Warp message should include:
+    // AddressedCall with Payload:
     //   - SubnetID
     //   - NodeID (must be Ed25519 NodeID)
-    //   - Weight of the validator
+    //   - Weight
     //   - Expiry
     //   - Ed25519 Signature over "[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]"
-    //   - BLS multisig over the above payload
     Message warp.Message `json:"message"`
 }
 ```
 
-After the `RegisterSubnetValidatorTx` is accepted on the P-Chain, the Subnet Validator is added to the Subnet's validator set. The P-Chain will then send a Warp message to the Subnet notifying of the validator set addition.
+The `Message` field must be an `AddressedCall` with the payload:
+
+```text
++-----------+----------+-----------+
+|   codecID :   uint16 |   2 bytes |
++-----------+----------+-----------+
+|    typeID :   uint32 |   4 bytes |
++-----------+----------+-----------+
+|  subnetID : [32]byte |  32 bytes |
++-----------+----------+-----------+
+|    nodeID : [32]byte |  32 bytes |
++-----------+----------+-----------+
+|    weight :   uint64 |   8 bytes |
++-----------+----------+-----------+
+|    expiry :   uint64 |   8 bytes |
++-----------+----------+-----------+
+| signature : [64]byte |  64 bytes |
++-----------+----------+-----------+
+                       | 150 bytes |
+                       +-----------+
+```
+
+- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
+- `typeID` is the payload type identifier and is `0x00000000` for this transaction
+- `subnetID`, `nodeID`, and `weight` are for the Subnet Validator being added
+- `expiry` is the time after which this message is invalid. After the P-Chain timestamp is past `expiry`, this Avalanche Warp Message can no longer be used to add the `nodeID` to the validator set of `subnetID`.
+
+    `validationID` is defined as the SHA256 hash of the `Payload` of the `AddressedCall`. This SHA256 hash will be used for replay protection. Used `validationID`s will be stored on the P-Chain. If a `RegisterSubnetValidatorTx`'s `validationID` has already been used, the transaction will be considered invalid. To prevent storing an unbounded number of `validationID`s, the `expiry` is required to be no longer than 48 hours in the future of the time the transaction is issued on the P-Chain. Any `validationIDs` with `expiry` more than 48 hours in the past can be flushed from the P-Chain's state.
+- `signature` is the raw bytes of the Ed25519 signature over the concatenated bytes of `[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]`. This signature must correspond to the Ed25519 public key that is used for the `nodeID`. This approach prevents NodeIDs from being unwillingly added to Subnets. `balance` is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64. A Subnet may choose to validate that the `signature` field is well-formed but it is not required.
```

Comment:
```suggestion
- `signature` is the raw bytes of the Ed25519 signature over the concatenated bytes of `[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]`. This signature must correspond to the Ed25519 public key that is used for the `nodeID`. This approach prevents NodeIDs from being unwillingly added to Subnets. `balance` is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64. A Subnet may choose to validate that the `signature` field is well-formed, but it is not required.
```

Should we move this bullet point above the `validationID` description? Seems a little odd in the middle of the field descriptions.

Reply:
Removed `signature` from warp message to use `blsPublicKey` directly.